src/QsCompiler/Transformations/CallGraph/ConcreteCallGraphWalker.cs

```diff
@@ -0,0 +1,275 @@
+                while (walker.SharedState.RequestStack.TryPop(out var currentRequest))
+                {
+                    // If there is a call to an unknown callable, throw exception
+                    if (!globals.TryGetValue(currentRequest.CallableName, out QsCallable currentCallable))
+                    {
+                        throw new ArgumentException($"Couldn't find definition for callable: {currentRequest.CallableName}");
+                    }
+
+                    var spec = GetSpecializationFromRequest(currentRequest, currentCallable);
+
+                    // The current request must be added before it is processed to prevent
+                    // self-references from duplicating on the stack.
+                    walker.SharedState.ResolvedNodeSet.Add(currentRequest);
+                    walker.SharedState.CurrentNode = currentRequest;
+                    walker.Namespaces.OnSpecializationImplementation(spec.Implementation);
+                }
+            }
+
+            /// <summary>
+            /// Returns the specialization that the request is referring to.
+            /// </summary>
+            private static QsSpecialization GetSpecializationFromRequest(ConcreteCallGraphNode request, QsCallable callable)
+            {
+                var relevantSpecs = callable.Specializations.Where(s => s.Kind == request.Kind);
+                var typeArgs = QsNullable<ImmutableArray<ResolvedType>>.Null;
+                if (callable.Signature.TypeParameters.Any())
+                {
+                    // Convert ParamResolutions from a dictionary to a type argument list
+                    typeArgs = QsNullable<ImmutableArray<ResolvedType>>.NewValue(callable.Signature.TypeParameters.Select(typeParam =>
+                    {
+                        if (typeParam is QsLocalSymbol.ValidName name)
+                        {
+                            if (request.ParamResolutions.TryGetValue(Tuple.Create(callable.FullName, name.Item), out var res))
+                            {
+                                return res;
+                            }
+                            else
+                            {
+                                throw new ArgumentException($"Couldn't resolve all type parameters for {callable.FullName}");
+                            }
+                        }
+                        else
+                        {
+                            throw new ArgumentException("Encountered invalid type parameter name during call graph construction");
+                        }
+                    }).ToImmutableArray());
+                    typeArgs = SpecializationBundleProperties.BundleId(typeArgs);
+                }
+                else if (relevantSpecs.Count() != 1)
+                {
+                    throw new ArgumentException($"Missing specialization {request.Kind} for {request.CallableName}");
```

Comment:
With this program:

```qsharp
function Foo() : Unit { }

@EntryPoint()
function Main() : Unit {
    let f = Foo;
    f();
}
```

I get:

> System.ArgumentException: Missing specialization QsControlledAdjoint for Graph.Foo

Reply:
I see the issue. When referencing a callable without calling it, the logic is overly cautious and assumes a link for all functor specializations. Since Foo does not have a body for all functor specializations, this causes an error. What it should do is assume a link for all functor specializations _defined for that callable_. I believe I can get this information from the callable, but I'll take a closer look in a bit.

Reply:
I've added some unit tests to cover these scenarios. They currently don't pass because I'm still working on the bug.

Reply:
fixed